### PR TITLE
bugfix: support more built-in parameters when set chash balancer

### LINKED ISSUE #717: bug: Can't set other Nginx built-in parameters when config chash balancer
### Environment

* apisix version (cmd: `0.8`):
* OS: Centos 7.4

it should support other built-in parameters like `uri,hostname`

### PR DESCRIPTION
**Pull request by lilien1010**
bugfix: support more built-in parameters when set chash balancer
fix: #717

 

### Summary

 support more `uri|server_name|server_addr|request_uri|remote_port|remote_addr|query_string|host|hostname)|arg_[0-9a-zA-z_-]+`

### Full changelog

* lua\apisix\schema_def.lua

### Issues resolved

Fix #717

### CODE REVIEW

File: t/admin/upstream.t [hunk 1]
```diff
@@ -782,3 +782,64 @@ GET /t
...
+            ngx.print(body)
+        }
+    }
+--- request
+GET /t
+--- error_code: 400
+--- response_body
+{"error_msg":"invalid configuration: invalid \"pattern\" in docuement at pointer \"#\/key\""}
+--- no_error_log
+[error]
```
Comment (membphis) [inline t/admin/upstream.t]: Need a blank line at the end of the file

**Comment by membphis** [comment]
add a new test case about `arg_xxx`, which is special.

File: lua/apisix/schema_def.lua [hunk 2]
```diff
@@ -236,7 +236,7 @@ local upstream_schema = {
         key = {
             description = "the key of chash for dynamic load balancing",
             type = "string",
-            enum = {"remote_addr"},
+            pattern = [[^((uri|server_name|server_addr|request_uri|remote_port|remote_addr|query_string|host|hostname)|arg_[0-9a-zA-z_-]+)$]],
```
Comment (moonming) [inline lua/apisix/schema_def.lua]: need add those info to document

File: t/admin/upstream.t [hunk 3]
```diff
@@ -782,3 +782,94 @@ GET /t
...
+
+
+
+=== TEST 24: set upstream with args(type: chash)
```
Comment (moonming) [inline t/admin/upstream.t]: we need more test cases, because you add many keys.

File: lua/apisix/schema_def.lua (hunk 2 above)
Comment (membphis) [inline lua/apisix/schema_def.lua]: code style: Single line code should be less than 80 characters long

File: t/admin/upstream.t [hunk 4]
```diff
@@ -782,3 +782,94 @@ GET /t
...
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/upstreams/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                    "key": "arg_device_id",
```
Comment (moonming) [inline t/admin/upstream.t]: Is this key realy works? you need add e2e(end to end) test cases for that.

Comment (lilien1010) [reply t/admin/upstream.t]: it  does works,`arg_device_id` means get `device_id` from  queryString

File: lua/apisix/schema_def.lua (hunk 2 above)
Comment (lilien1010) [reply lua/apisix/schema_def.lua]: got it

Comment (lilien1010) [reply lua/apisix/schema_def.lua]: I can modify the documents later

File: t/admin/upstream.t (hunk 4 above)
Comment (moonming) [reply t/admin/upstream.t]: you not got my point. 
You set the rule successfully, however, will the request be processed according to the rules you set?

Comment (moonming) [reply t/admin/upstream.t]: for example: https://github.com/iresty/apisix/blob/master/t/node/chash-balance.t

File: lua/apisix/schema_def.lua [hunk 5]
```diff
@@ -236,7 +236,8 @@ local upstream_schema = {
         key = {
             description = "the key of chash for dynamic load balancing",
             type = "string",
-            enum = {"remote_addr"},
+            pattern = [[^((uri|server_name|server_addr|request_uri|remote_port]]
+            ..[[|remote_addr|query_string|host|hostname)|arg_[0-9a-zA-z_-]+)$]],
```
Comment (membphis) [inline lua/apisix/schema_def.lua]: bad indentation

```lua
            pattern = [[^((uri|server_name|server_addr|request_uri|remote_port]]
                      ..[[|remote_addr|query_string|host|hostname)|arg_[0-9a-zA-z_-]+)$]],
```

File: t/admin/upstream.t (hunk 4 above)
Comment (membphis) [reply t/admin/upstream.t]: I think we can write some debug log to print the `key`: 

https://github.com/iresty/apisix/blob/master/lua/apisix/balancer.lua#L158

and then we can check the content of error log to confirm if the balancer works fine with new `key`.

**Comment by membphis** [comment]
we need to fix this test case.

```
#   Failed test 'TEST 25:  wrong upstream key - response_body - response is expected (repeated req 0, req 0)'
#   at /home/travis/build/iresty/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 1589.
#          got: '{"error_msg":"invalid configuration: property \"key\" validation failed: failed to match pattern \"^((uri|server_name|server_addr|request_uri|remote_port|remote_addr|query_string|host|hostname)|arg_[0-9a-zA-z_-]+)$\" with \"not_support\""}
# '
#     expected: '{"error_msg":"invalid configuration: invalid \"pattern\" in docuement at pointer \"#\/key\""}
# '
t/admin/upstream.t ..................... 74/? # Looks like you failed 1 test of 76.
t/admin/upstream.t ..................... Dubious, test returned 1 (wstat 256, 0x100)
```

**Comment by moonming** [comment]
@lilien1010 are issues in comments fixed? please press 'Resolve conversation' after you fixed.

File: doc/architecture-design.md [hunk 6]
```diff
@@ -211,7 +211,7 @@ In addition to the basic complex equalization algorithm selection, APISIX's Upst
...
 |type            |required|`roundrobin` supports the weight of the load, `chash` consistency hash, pick one of them.|
 |nodes           |required|Hash table, the key of the internal element is the upstream machine address list, the format is `Address + Port`, where the address part can be IP or domain name, such as `192.168.1.100:80`, `foo.com:80`, etc. Value is the weight of the node. In particular, when the weight value is `0`, it has a special meaning, which usually means that the upstream node is invalid and never wants to be selected.|
-|key             |required|This option is only valid if the type is `chash`. Find the corresponding node `id` according to `key`, the same `key` in the same object, always return the same id.|
+|key             |required|This option is only valid if the type is `chash`. Find the corresponding node `id` according to `key`, the same `key` in the same object, always return the same id. For now, it support nginx built-in variables like `uri,server_name,server_addr,request_uri,remote_port,remote_addr,query_string,host,hostname,arg_***` , `arg_***` is arguments in the request line, [Index of Nginx variables](http://nginx.org/en/docs/varindex.html)|
```
Comment (moonming) [inline doc/architecture-design.md]: please add one space after `,`

**Comment by lilien1010** [comment]
it wired, I already rebased from remote master, and made the file comparision with 'beyond compare' ,it still doesn't  pass the CI

**Comment by juzhiyuan** [comment]
@lilien1010 because the current branch doesn't have the latest Dashboard codes, we may need update the dashboard submodule

**Comment by lilien1010** [comment]
@juzhiyuan 
I am stuck on the problem, would you give me some advice how can I fix that
here is my repo https://github.com/lilien1010/apisix/tree/patch-3

**Comment by moonming** [comment]
> @juzhiyuan
> I am stuck on the problem, would you give me some advice how can I fix that
> here is my repo https://github.com/lilien1010/apisix/tree/patch-3

I think you have two choices:
1. rollback commit about dashboard;
2. rebase master. If not works, then you have to do choice one.

File: doc/architecture-design.md (hunk 6 above)
Comment (lilien1010) [reply doc/architecture-design.md]: changed

**Comment by moonming** [comment]
I checked your commits, are you rebased the master? 
 I think to close this PR and create a new one from the master is easier.

**Comment by lilien1010** [comment]
close this commit due to too many error commits,
will be better for the next PR

**Pull request closed without merging**